action/candidate_update.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+	"github.com/iotexproject/iotex-core/pkg/version"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+)
+
+// CandidateUpdate is the action to register a candidate
+type CandidateUpdate struct {
+	AbstractAction
+
+	name            string
+	operatorAddress address.Address
+	rewardAddress   address.Address
+}
+
+// NewCandidateUpdate creates a CandidateUpdate instance
+func NewCandidateUpdate(
+	nonce uint64,
+	name, operatorAddrStr, rewardAddrStr string,
+	gasLimit uint64,
+	gasPrice *big.Int,
+) (*CandidateUpdate, error) {
+	cu := &CandidateUpdate{
+		AbstractAction: AbstractAction{
+			version:  version.ProtocolVersion,
+			nonce:    nonce,
+			gasLimit: gasLimit,
+			gasPrice: gasPrice,
+		},
+		name: name,
+	}
+
+	var err error
+	if len(operatorAddrStr) > 0 {
+		cu.operatorAddress, err = address.FromString(operatorAddrStr)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(rewardAddrStr) > 0 {
+		cu.rewardAddress, err = address.FromString(rewardAddrStr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return cu, nil
+}
+
+// Name returns candidate name to update
+func (cu *CandidateUpdate) Name() string { return cu.name }
+
+// OperatorAddress returns candidate operatorAddress to update
+func (cu *CandidateUpdate) OperatorAddress() address.Address { return cu.operatorAddress }
+
+// RewardAddress returns candidate rewardAddress to update
+func (cu *CandidateUpdate) RewardAddress() address.Address { return cu.rewardAddress }
+
+// Serialize returns a raw byte stream of the CandidateUpdate struct
+func (cu *CandidateUpdate) Serialize() []byte {
+	return byteutil.Must(proto.Marshal(cu.Proto()))
+}
+
+// Proto converts to protobuf CandidateUpdate Action
+func (cu *CandidateUpdate) Proto() *iotextypes.CandidateBasicInfo {
+	act := &iotextypes.CandidateBasicInfo{}
+
+	if len(cu.name) > 0 {
+		act.Name = cu.name
+	}
+
+	if cu.operatorAddress != nil {
+		act.OperatorAddress = cu.operatorAddress.String()
+	}
+
+	if cu.rewardAddress != nil {
+		act.RewardAddress = cu.rewardAddress.String()
+	}
+
+	return act
+}
+
+// LoadProto converts a protobuf's Action to CandidateUpdate
+func (cu *CandidateUpdate) LoadProto(pbAct *iotextypes.CandidateBasicInfo) error {
+	if pbAct == nil {
+		return errors.New("empty action proto to load")
+	}
+
+	cu.name = pbAct.GetName()
+
+	if len(pbAct.GetOperatorAddress()) > 0 {
+		operatorAddr, err := address.FromString(pbAct.GetOperatorAddress())
+		if err != nil {
+			return err
+		}
+		cu.operatorAddress = operatorAddr
+	}
+
+	if len(pbAct.GetRewardAddress()) > 0 {
+		rewardAddr, err := address.FromString(pbAct.GetRewardAddress())
+		if err != nil {
+			return err
+		}
+		cu.rewardAddress = rewardAddr
+	}
+	return nil
+}
+
+// IntrinsicGas returns the intrinsic gas of a CandidateUpdate
+func (cu *CandidateUpdate) IntrinsicGas() (uint64, error) {
+	return 0, nil
```

Comment:
is this a TODO?

Reply:
Referenced from other actions without payload. intrinsicgas is just set to 0